README.md
### EKS CLI wrapper

[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)](https://github.com/pre-commit/pre-commit)

#### Synopsis

```shell
$ ./cluster.sh -h
Usage: ./cluster.sh { -h help | -c create a cluster | -d dry run | -D delete a cluster | -a add node group to existing cluster | -r remove node group from existing cluster | -g get kubeconfig command | -s show cluster and cluster nodegroups }
```

This tool is meant to aid Support/SEs in creating and deleting (VERY simple) EKS clusters for testing purposes. Most flags are left to the defaults as of this release.

> **WARNING:** do not put spaces, underscores (`_`) or periods (`.`) in any of these input fields (dashes `-` are okay, however). Chances are, it will end up costing you a lot of time before `eksctl` errors out.

If you wish to add additional configuration options (cf. `eksctl -h`, or `eksctl create cluster -h` for further
information), you may pass a `-d` (dry run) flag to the script to dump the command to `stdout` for further
modification or review.
```shell
$ ./cluster.sh -d
Cluster name (optional): brandon-test
Your name (for tagging purposes, defaults to `Support`): Brandon
Override K8s version (defaults to 1.18):
WARNING: Defaulting to 1.18
Cluster region (default `us-east-2`):
WARNING: Defaulting to us-east-2
Min number of worker nodes (default is 2):
WARNING: Defaulting to 2
Max number of worker nodes (default is 2 + 1 = 3):
WARNING: Defaulting to 3
Instance size (default is `t3.medium`):
WARNING: Defaulting to t3.medium
Instance volume size (default 80 (GB)):
WARNING: Defaulting to 80
Nodegroup name (default `test`, no duplicates):
WARNING: Defaulting to test
Nodegroup labels (default `env=test`; must be in format key1=val1,key2=val2): env=test,agent=2-3-1

Dry run output:

eksctl create cluster --name brandon-test --tags "Owner=Brandon" --region us-east-2 --version 1.18 --timeout "60m0s" --nodegroup-name test --node-type t3.medium --nodes-min 2 --nodes-max 3 --node-volume-size 80 --node-labels "env=test,agent=2-3-1" --full-ecr-access --asg-access --alb-ingress-access --auto-kubeconfig

```

#### Prerequisites

This script is intended to be executed with `bash` and requires `ossp-uuid` (Mac, or `uuid` on Linux), a working `eksctl` installation (>= v0.31.0), and the AWS CLI. Refer to the [AWS
docs](https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html) to install the latter utilities.

#### Examples

Because output can be rather extensive, I've opted to link to various examples to take you through a cluster's lifecycle ~

* [Creating a cluster](examples/create_cluster.md)
* [Adding a nodegroup](examples/add_nodegroup.md)
* [Deploying an application](examples/deploy_application.md)
* [Removing a nodegroup](examples/remove_nodegroup.md)
* [Deleting a cluster](examples/delete_cluster.md)

#### Updates

See the [Changelog](CHANGELOG.md) for fixes and updates by version.


CHANGELOG.md
## Changelog

This changelog notes important changes with each minor release.

### v1.0.0

*08/08/2021*

* Added automations to generate releases and docker image builds.

### v0.4.1

*02/08/2021*

* Fixed synopsis in README.

### v0.4

*02/08/2021*

* Resolved the fact that later versions of `eksctl` seem to add additional verbosity on versioning.
* Fixed `Usage:` string to show that only one flag can be passed in at a time.
* Minor code readability fixes.

### v0.3

*02/07/2021*

* Found that the problem with gathering node groups on clusters that don't currently have any was fixed in a later
  `eksctl` version than 0.31.0 (tested with v0.37.0, the latest). Will keep these fixes in-place, however, for those
  using an earlier `eksctl` version.
* Fixed the portion about adding additional nodegroups following cluster creation. Was formerly broken. Now you may add as many nodegroups as you would like while building a cluster.

### v0.2

*02/06/2021*

* Ability to add and remove nodegroups.
* Dump the AWS CLI `update-kubeconfig` command to connect to your cluster (must be ran as root).
* Show cluster details, including attached nodegroups.
* Various fixes to make the interface more user-friendly.

#### TODOs following this release:
* Identified a problem with gathering node groups on clusters that don't currently have any (cf. `FIXME`s in `eks.sh`). This could be fixed in a later version of `eksctl`, however. I am testing with `v0.31.0`.
* I need to fix the portion about adding additional nodegroups following cluster creation; currently broken.

### v0.1

*02/05/2021*

* Basic EKS cluster creation and deletion in AWS regions.


cluster.sh
#! /bin/bash
# Stand up a K8s cluster with eksctl.


source src/utils.sh
source src/eks.sh


usage()
{
    printf "Usage: %s { -h help | -c create a cluster | -d dry run | -D delete a cluster \
| -a add node group to existing cluster | -r remove node group from existing cluster \
| -g get kubeconfig command | -s show cluster and cluster nodegroups }\\n" "$0" >&2
    exit 1
}


if [ $# -ne 1 ]
then
    printf "Expected 1 argument to $0, received %s.\\n" "$#" >&2
    usage
    exit 1
fi


while getopts ":h:dDcagrs" flag
do
    case "${flag}" in
        d)
            # Dry run.
            create true
            ;;
        c)
            # Create a cluster (w/o a dry run output).
            create false
            ;;
        D)
            delete
            ;;
        a)
            add_nodes
            ;;
        r)
            remove_nodes
            ;;
        g)
            get_kubeconfig
            ;;
        s)
            show_cluster
            ;;
        h)
            # Help.
            usage
            ;;
        *)
            usage
            ;;
    esac
done


src/utils.sh
#! /bin/bash

##
# Print an error message to stderr.
_error()
{
    if [ $# -ne 1 ]
    then
        printf "Expected 1 argument to \`_error\`, received %s.\\n" "$#" >&2
        exit 1
    fi

    local message
    message="$1"

    printf "\e[2m\e[1mERROR\e[0m\e[2m: %s\e[0m\\n" "$message" >&2
}


##
# Print a warning message to stderr.
_warning()
{
    if [ $# -ne 1 ]
    then
        _error "Expected 1 argument to \`_warning\`, received $#.\\n"
    fi

    local message
    message="$1"

    printf "\e[2m\e[1mWARNING\e[0m\e[2m: %s\e[0m\\n" "$message" >&2
}


##
# Print the equivalent of a white-text header/separator.
_separator()
{
    if [ $# -ne 1 ]
    then
        _error "Expected 1 argument to \`_separator\`, received $#.\\n"
    fi

    local message
    message="$1"

    printf "\\n\e[1m%s\e[0m\\n\\n" "$message"

    return 0
}


_wait()
{
    if [ $# -ne 1 ]
    then
        printf "Expected 1 argument to \`_wait\`, received %s.\\n" "$#"
        exit 1
    fi

    local wait_time
    wait_time="$1"

    for _ in $(seq 1 "$wait_time")
    do
        printf "..."
        sleep 1
    done
    printf " should be good to go now!\\n"
}


##
# Get a user's response on a question to set environment variables.
response()
{
    if [ $# -eq 0 ]
    then
        _error "Must submit at least 2 arguments to \`response\` function for IO."
        exit 1
    elif [ $# -gt 2 ]
    then
        _warning "received >2 arguments at response function, ignoring extra arguments"
    fi

    question="$1"
    default="$2"

    read -r -p "$question" var
    if [ "$var" ]
    then
        printf "%s" "$var"
    else
        if [ "$default" ]
        then
            _warning "Defaulting to $default"
        else
            _warning "Attempted to default, but no value given, returning \"\""
        fi
        printf "%s" "$default"
    fi

    return 0
}


create_cluster.sh
#! /bin/bash
# Stand up a K8s cluster with eksctl.


source src/utils.sh


usage()
{
    printf "Usage: %s [-h help] [-d dry run] [-D delete a cluster]\\n" "$0"
    exit 1
}


_dry_run=false

while getopts ":h:dD" flag
do
    case "${flag}" in
        d)
            # Dry run.
            _dry_run=true
            ;;
        D)
            # Delete a cluster.
            REGION="$(response "Cluster region (defaults to \`us-east-2\`): " "us-east-2")"
            list="$(eksctl get cluster --region "$REGION")"

            if [[ "$list" =~ "No clusters found" ]]
            then
                _error "$list"
                exit 1
            else
                printf "%s\\n" "$list"
            fi

            CLUSTER="$(response "Select a cluster from the above list to delete: " "")"

            if [ "$CLUSTER" ]
            then
                eksctl delete cluster --region "$REGION" "$CLUSTER"
            else
                _error "Must provide a cluster name."
            fi

            exit 1
            ;;
        h)
            # Help.
            usage
            ;;
        *)
            usage
            ;;
    esac
done


# Cluster
NAME="$(response "Cluster name (optional): " "support-test-$(uuid)")"
OWNER="$(response "Your name (for tagging purposes, defaults to \`Support\`): " "Support")"
VERSION="$(response "Override K8s version (defaults to 1.18): " "1.18")"
REGION="$(response "Cluster region (default \`us-east-2\`): " "us-east-2")"


##
# Get Nodegroup values from the user.
nodegroup_io()
{
    MIN="$(response "Min number of worker nodes (default is 2): " "2")"
    MAX="$(response "Max number of worker nodes (default is $MIN + 1 = $(( MIN + 1 ))): " "$(( MIN + 1 ))")"
    SIZE="$(response "Instance size (default is \`t3.medium\`): " "t3.medium")"
    EBS_SIZE="$(response "Instance volume size (default 80 (GB)): " "80")"
    NODEGROUP_NAME="$(response "Nodegroup name (default \`test\`): " "test")"

    # TODO: Add check/retry to ensure there are no duplicate nodegroup names.
}


main()
{
    # Set nodegroup variables prior to creating a cluster for the first time.
    nodegroup_io

    # Create
    while true
    do
        cmd="eksctl create cluster --name ${NAME} --tags \"Owner=${OWNER}\" --region ${REGION} \
--version ${VERSION} --timeout \"60m0s\" --nodegroup-name ${NODEGROUP_NAME} --node-type ${SIZE} \
--nodes-min ${MIN} --nodes-max ${MAX} --node-volume-size ${EBS_SIZE} --node-labels \"env=test\" \
--full-ecr-access --asg-access --alb-ingress-access --auto-kubeconfig"

        if $_dry_run
        then
            _separator "Dry run output:"
            printf "%s\\n\\n" "$cmd"
            exit 0
        fi

        _separator "Creating EKS Cluster"

        eval "$cmd"

        if ! [ $? ]
        then
            # Cluster creation was not successful, so you can either retry following its complete deletion, or exit.
            printf "Deleting cluster and pausing\\n"

            # This command blocks, but it doesn't exit when all resources are deleted, yet.
            eksctl delete cluster --region "$REGION" "$NAME"

            _wait 90

            retry="$(response "Retry? [y/n]: ")"
            if [[ ! "$retry" =~ ^[Yy].* ]]
            then
                return 1
            fi
        else
            break
        fi
    done

    _separator "EKS Cluster $NAME created. Moving on to optional additional compute"
    printf "\\n"

    # Cluster creation was successful. Want to add more node(s|groups)?
    while true
    do
        if [[ "$(response "Would you like to add further nodegroups? [y/n]: " "no")" =~ ^[Yy].* ]]
        then
            # Yes, so reset nodegroup variables and run an additional command to append a nodegroup.
            _separator "Adding additional nodegroup to cluster $NAME"
            nodegroup_io
            :
        else
            # No, let's just exit with some beautiful output :D
            :
            break
        fi
    done
}


main

src/eks.sh
#! /bin/bash

##
# Delete a cluster from a menu.
delete()
{
    # Delete a cluster.
    REGION="$(response "Cluster region (defaults to \`us-east-2\`): " "us-east-2")"
    list="$(eksctl --verbose 2 get cluster --region "$REGION")"

    if [ -z "$list" ]
    then
        # Above command probably sent text to stderr we didn't capture.
        exit 1
    elif [[ "$list" =~ "No clusters found" ]]
    then
        _error "$list"
        exit 1
    else
        printf "%s\\n" "$list"
    fi

    NAME="$(response "Select a cluster from the above list to delete: " "")"

    if [ "$NAME" ]
    then
        _separator "Starting delete of cluster $NAME"
        eksctl delete cluster --region "$REGION" "$NAME"
        _wait 60
    else
        _error "Must provide a cluster name."
    fi

    return 0
}


##
# Get Nodegroup values from the user.
_nodegroup_io()
{
    MIN="$(response "Min number of worker nodes (default is 2): " "2")"
    MAX="$(response "Max number of worker nodes (default is $MIN + 1 = $(( MIN + 1 ))): " "$(( MIN + 1 ))")"
    SIZE="$(response "Instance size (default is \`t3.medium\`): " "t3.medium")"
    EBS_SIZE="$(response "Instance volume size (default 80 (GB)): " "80")"
    NODEGROUP_NAME="$(response "Nodegroup name (default \`test\`, no duplicates): " "test")"
    LABELS="$(response "Nodegroup labels (default \`env=test\`; must be in format key1=val1,key2=val2): " "env=test")"

    return 0
}


##
# Add a nodegroup to a cluster.
add_nodes()
{
    if [ $# -eq 1 ]
    then
        # This method is being called upon a cluster that was just created, so we already have the cluster info.
        :
    else
        # We do not have the cluster info, so necessary variables are not set.
        REGION="$(response "Cluster region (defaults to \`us-east-2\`): " "us-east-2")"
        list="$(eksctl --verbose 2 get cluster --region "$REGION")"

        if [ -z "$list" ]
        then
            exit 1
        elif [[ "$list" =~ "No clusters found" ]]
        then
            _error "$list"
            exit 1
        else
            printf "%s\\n" "$list"
        fi

        NAME="$(response "Select a cluster from the above list to add a node group to: " "")"
        while ! [ "$NAME" ]
        do
            _error "Expected cluster name."
            NAME="$(response "Select a cluster from the above list to add a node group to: " "")"
        done

        OWNER="$(response "Your name (for tagging purposes, defaults to \`Support\`): " "Support")"
    fi

    # This is a bug in `eksctl`. If no nodegroup exists, it prints "Error: Nodegroup with name  not found" to stderr
    # in earlier eksctl versions. Was fixed somewhere between v0.31.0 and v0.37.0.
    list="$(eksctl --verbose 2 get nodegroup --region "$REGION" --cluster "$NAME" 2>&1)"
    if [[ "$list" =~ "Error:" ]]
    then
        _warning "There are currently no nodegroups attached to this cluster"
    else
        printf "%s\\n" "$list"
    fi

    _nodegroup_io
    _separator "Creating EKS Nodegroup"

    eksctl create nodegroup --cluster "$NAME" --tags "\"Owner=$OWNER\"" --region "$REGION" \
--timeout "60m0s" --name "$NODEGROUP_NAME" --node-type "$SIZE" --nodes-min "$MIN" \
--nodes-max "$MAX" --node-volume-size "$EBS_SIZE" --node-labels "${LABELS}" --full-ecr-access \
--asg-access --alb-ingress-access

    printf "\\n"
    eksctl --verbose 2 get nodegroup --region "$REGION" --cluster "$NAME"
    printf "\\n"

    return 0
}


##
# Remove node group from an existing cluster.
remove_nodes()
{
    # We do not have the cluster info, so necessary variables are not set.
    REGION="$(response "Cluster region (defaults to \`us-east-2\`): " "us-east-2")"
    list="$(eksctl --verbose 2 get cluster --region "$REGION")"

    if [ -z "$list" ]
    then
        exit 1
    elif [[ "$list" =~ "No clusters found" ]]
    then
        _error "$list"
        exit 1
    else
        printf "%s\\n" "$list"
    fi

    NAME="$(response "Select a cluster from the above list to remove a node group from: " "")"
    if ! [ "$NAME" ]
    then
        _error "Expected cluster name."
    fi

    # This is a bug in `eksctl`. If no nodegroup exists, it prints "Error: Nodegroup with name  not found" to stderr
    # in earlier eksctl versions. Was fixed somewhere between v0.31.0 and v0.37.0.
    list="$(eksctl --verbose 2 get nodegroup --region "$REGION" --cluster "$NAME" 2>&1)"

    if [ -z "$list" ]
    then
        # Above command probably sent text to stderr we didn't capture.
        exit 1
    elif [[ "$list" =~ "Error:" ]]
    then
        _error "$list"
        exit 1
    else
        printf "%s\\n" "$list"
    fi

    NODEGROUP_NAME="$(response "Select a nodegroup (name) from the above list to remove from cluster $NAME: " "")"

    _separator "Deleting nodegroup $NODEGROUP_NAME from cluster $NAME"

    eksctl delete nodegroup --region "$REGION" --cluster "$NAME" --name "$NODEGROUP_NAME"
    printf "\\n"

    return 0
}


##
# Create an EKS cluster.
create()
{
    if [ $# -ne 1 ]
    then
        _error "Expected 1 argument to \`create\`, received $#.\\n"
    fi

    _dry_run=$1

    # Cluster
    NAME="$(response "Cluster name (optional): " "support-test-$(uuid)")"
    OWNER="$(response "Your name (for tagging purposes, defaults to \`Support\`): " "Support")"
    VERSION="$(response "Override K8s version (defaults to 1.18): " "1.18")"
    REGION="$(response "Cluster region (default \`us-east-2\`): " "us-east-2")"

    # Set nodegroup variables prior to creating a cluster for the first time.
    _nodegroup_io

    # Create
    while true
    do
        cmd="eksctl create cluster --name ${NAME} --tags \"Owner=${OWNER}\" --region ${REGION} \
--version ${VERSION} --timeout \"60m0s\" --nodegroup-name ${NODEGROUP_NAME} --node-type ${SIZE} \
--nodes-min ${MIN} --nodes-max ${MAX} --node-volume-size ${EBS_SIZE} --node-labels \"${LABELS}\" \
--full-ecr-access --asg-access --alb-ingress-access --auto-kubeconfig"

        if $_dry_run
        then
            _separator "Dry run output:"
            printf "%s\\n\\n" "$cmd"
            exit 0
        fi

        _separator "Creating EKS Cluster"

        eval "$cmd"

        if ! [ $? ]
        then
            # Cluster creation was not successful, so you can either retry following its complete deletion, or exit.
            printf "Deleting cluster and pausing\\n"

            # This command blocks, but it doesn't exit when all resources are deleted, yet.
            eksctl delete cluster --region "$REGION" "$NAME"

            _wait 60

            retry="$(response "Retry? [y/n]: ")"
            if [[ ! "$retry" =~ ^[Yy].* ]]
            then
                return 1
            fi
        else
            break
        fi
    done

    _separator "EKS Cluster $NAME created. Moving on to optional additional compute"

    # Cluster creation was successful. Want to add more node(s|groups)?
    while [[ "$(response "Would you like to add further nodegroups? [y/n]: " "no")" =~ ^[Yy].* ]]
    do
        # Yes, so reset nodegroup variables and run an additional command to append a nodegroup.
        _separator "Adding additional nodegroup to cluster $NAME"
        add_nodes "$NAME"
    done

    _separator "Execute the following command to connect to your new cluster:"
    printf "\$ sudo aws eks --region %s update-kubeconfig --name %s\\n\\n" "$REGION" "$NAME"

    return 0
}


##
# Dump command you need to execute as root to connect ot a cluster.
get_kubeconfig()
{
    REGION="$(response "Cluster region (defaults to \`us-east-2\`): " "us-east-2")"
    list="$(eksctl --verbose 2 get cluster --region "$REGION")"

    if [ -z "$list" ]
    then
        exit 1
    elif [[ "$list" =~ "No clusters found" ]]
    then
        _error "$list"
        exit 1
    else
        printf "%s\\n" "$list"
    fi

    NAME="$(response "Select a cluster from the above list to generate \`update-kubeconfig\` command on: " "")"

    if ! [ "$NAME" ]
    then
        _error "Expected cluster name."
    fi

    _separator "Execute the following command to connect to your new cluster:"
    printf "\$ sudo aws eks --region %s update-kubeconfig --name %s\\n\\n" "$REGION" "$NAME"

    return 0
}


##
# Show information about a cluster.
show_cluster()
{
    REGION="$(response "Cluster region (defaults to \`us-east-2\`): " "us-east-2")"
    list="$(eksctl --verbose 2 get cluster --region "$REGION")"

    if [ -z "$list" ]
    then
        exit 1
    elif [[ "$list" =~ "No clusters found" ]]
    then
        _error "$list"
        exit 1
    else
        printf "%s\\n" "$list"
    fi

    NAME="$(response "Select a cluster from the above list to show info: " "")"

    if ! [ "$NAME" ]
    then
        _error "Expected cluster name."
    fi

    # This is a bug in `eksctl`. If no nodegroup exists, it prints "Error: Nodegroup with name  not found" to stderr
    # in earlier eksctl versions. Was fixed somewhere between v0.31.0 and v0.37.0.
    list="$(eksctl --verbose 2 get nodegroup --region "$REGION" --cluster "$NAME" 2>&1)"
    if [[ "$list" =~ "Error:" ]]
    then
        _warning "There are currently no nodegroups attached to this cluster"
    else
        printf "\\n%s\\n\\n" "$list"
    fi

    return 0
}
